File: accelerometer/mpu9250.cpp
#include "mpu9250.h";

mpu9250::mpu9250(){
    Wire.begin();
    // Set accelerometers low pass filter at 5Hz
    I2CwriteByte(MPU9250_ADDRESS,29,0x06);
    // Set gyroscope low pass filter at 5Hz
    I2CwriteByte(MPU9250_ADDRESS,26,0x06);
    // Configure gyroscope range
    I2CwriteByte(MPU9250_ADDRESS,27,GYRO_FULL_SCALE_1000_DPS);
    // Configure accelerometers range
    I2CwriteByte(MPU9250_ADDRESS,28,ACC_FULL_SCALE_4_G);
    // Set by pass mode for the magnetometers
    I2CwriteByte(MPU9250_ADDRESS,0x37,0x02);
    // Request continuous magnetometer measurements in 16 bits
    I2CwriteByte(MAG_ADDRESS,0x0A,0x16);
}

void mpu9250::readAccelerometer(){
    accelermeter.x = -(Buf[0]<<8 | Buf[1]);
    accelermeter.y = -(Buf[2]<<8 | Buf[3]);
    accelermeter.z = Buf[4]<<8 | Buf[5];
}

void mpu9250::readGyroscope(){
    gyroscope.x = -(Buf[8]<<8 | Buf[9]); 
    gyroscope.y = -(Buf[10]<<8 | Buf[11]);
    gyroscope.z = Buf[12]<<8 | Buf[13];
}

void mpu9250::readMagnetometer(){
    megnetometer.x=-(Mag[3]<<8 | Mag[2]);
    megnetometer.y=-(Mag[1]<<8 | Mag[0]);
    megnetometer.z=-(Mag[5]<<8 | Mag[4]);
}

void mpu9250::updateMagnetometerData(){
    uint8_t ST1;
    do{
        I2Cread(MAG_ADDRESS,0x02,1,&ST1);
    }
    while (!(ST1&0x01));
    I2Cread(MAG_ADDRESS,0x03,7,Mag);
}

void mpu9250::updateGyroscopeAndAccelerometerData(){
    I2Cread(MPU9250_ADDRESS,0x3B,14,Buf);
    readAccelerometer();
    readGyroscope();
}

vec3 mpu9250::getAccelerometerData(){
    return accelermeter;
}

vec3 mpu9250::getGyroscopeData(){
    return gyroscope;
}

vec3 mpu9250::getMagnetometerData(){
    return accelermeter;
}

mpu9250::~mpu9250(){
}

mpu9250 mpu = mpu9250();

File: raspberry/mpu9250.py
from smbus import SMBus

MPUADDRESS = 0x68
MAGADDRESS = 0x0c

class vec3: 
    def __init__(self):
        self.x = 0
        self.y = 0
        self.z = 0
    
    def __str__(self):
        return "{},{},{}\n".format(self.x, self.y, self.z)
    
    def __repr__(self):
        return "{},{},{}".format(self.x, self.y, self.z)


class mpu9250:
    def __init__(self):
        self.accelerometer = vec3()
        self.gyroscope = vec3()
        self.bus = SMBus(1)
        self.bus.open(1)
        # Set accelerometers low pass filter at 5Hz
        self.bus.write_byte_data(MPUADDRESS,29,6)
        # Set gyroscope low pass filter at 5Hz
        self.bus.write_byte_data(MPUADDRESS,26,6)
        # Configure gyroscope range
        self.bus.write_byte_data(MPUADDRESS,27,10)
        # Configure accelerometers range
        self.bus.write_byte_data(MPUADDRESS,28,8)
        # Set by pass mode for the magnetometers
        self.bus.write_byte_data(MPUADDRESS,37,2)
        # Request continuous magnetometer measurements in 16 bits
        #self.bus.write_byte_data(MAGADDRESS,0x0A,0x16)

    def updateGyroscopeAndAccelerometerData(self):
        data = self.bus.read_i2c_block_data(MPUADDRESS, 0x3b, 14)
        self.accelerometer.x =-(data[0]<<8 | data[1])
        self.accelerometer.y = -(data[2]<<8 | data[3])
        self.accelerometer.z = data[4]<<8 | data[5]
        self.gyroscope.x = -(data[8]<<8 | data[9])
        self.gyroscope.y = -(data[10]<<8 | data[11])
        self.gyroscope.z = data[12]<<8 | data[13]
